Add `LDF_PREFER_EAGER_ALLOCATIONS` for legate investigations

seberg opened the pull request:
Legate folks asked a way to use eager allocations more.  This might actually be useful for in some places (or at least the machinery around it).

I didn't set up the tests, I ran them locally with 2 GPUs and they run fine except, that using this for joins is actually incorrect so one tests (correctly) fails.

To use just run with `LDF_PREFER_EAGER_ALLOCATIONS=1`, and many ops will use eager allocations:
* to/from timestamp
* casts
* unary and binary ops
* replace_nulls
* Join...
  * If LEFT join
  * and explicit right side broadcasting
* ~parquet reading~ (removed in this PR for now)

(As said, even then it is incorrect: a left join can return a different sized result if there are duplicates in the RHS.  Doing this will be flagged as an error, and can lead to segfaults.)

---

I would be OK with putting this in if it helps for whatever tries someone wants to do.  I mean it if I say that the general ability may actually be useful even if the environment variable is clearly wrong with applying it to join.
It does add a bit of cruft, though.

RAMitchell commented on cpp/src/core/column.cu:
```diff
@@ -142,7 +142,9 @@ T* maybe_bind_buffer(legate::PhysicalStore store, std::size_t size)
-    out = store.write_accessor<T, 1>().ptr(0);
+    auto acc = store.write_accessor<T, 1>();
+    assert(store.shape<1>().hi[0] == -1 || acc.accessor.is_dense_row_major(store.shape<1>()));
+    out = acc.ptr(store.shape<1>().lo[0]);
```
Yes looks like a bug.

RAMitchell commented on cpp/include/legate_dataframe/core/column.hpp:
```diff
@@ -253,12 +271,22 @@ class LogicalColumn {
    * @return The new unbounded column
    */
-  static LogicalColumn empty_like(cudf::data_type dtype, bool nullable, bool scalar = false)
+  static LogicalColumn empty_like(cudf::data_type dtype,
```
Do we need all of these empty_like constructors?

RAMitchell commented on cpp/include/legate_dataframe/core/column.hpp:
```diff
@@ -603,27 +631,33 @@ class PhysicalColumn {
   /**
    * @brief Makes the unbound column empty. Valid only when the column is unbound.
+   *
+   * @param allow_copy If true, no error is given if the column is bound, this becomes
```
This comment is more descriptive than the above, maybe duplicate it there.

RAMitchell commented on cpp/include/legate_dataframe/core/column.hpp:
```diff
@@ -603,27 +631,33 @@ class PhysicalColumn {
+   * @param allow_copy If true, copies the data if the column is bound.
    */
-  void move_into(std::unique_ptr<cudf::column> column);
+  void move_into(std::unique_ptr<cudf::column> column, bool allow_copy = false);
```
So move into will lazily try to move the input if it can, and otherwise copy into any existing buffer?

Is there any argument for having a move_into and copy_into function to make this explicit?

seberg replied:
Yeah, you are right, that is nicer API wise.  I tried to avoid branching in the tasks for now.  But I guess a:
```
if get_():
    tbl.copy_into();
else:
    tbl.move_into();
```
at the end of a few tasks isn't bad since it is likely not quite permanent anyway.

RAMitchell commented on cpp/src/core/library.cpp:
```diff
@@ -153,4 +157,13 @@ legate::Library& get_library()
   return library;
 }
 
+bool get_prefer_eager_allocations()
```
This is not intended to be a long term solution right?

RAMitchell replied:
IMO, from a user perspective, users shouldn't to have to touch any parameters for streaming to work - it should just work. This is a demo so its fine for now.

RAMitchell reviewed (approved):
If I understand correctly, using eager allocations comes at the cost of extra memory, as the cudf/arrow result cannot be reused by moving into the legate frame.

For something like a binary op we could hypothetically implment it ourselves without arrow/cudf and have the best of both worlds. Just a thought.

seberg commented:
@RAMitchell thanks for the review. I added `copy_into()` and `move_into()`.  Unfortunately that (and a bit removing the `empty_like()` lead to some follow-ups and slight re-org for the `copy_into()`/`move_into()` branching:
* Tweak the definition of `unbound()`, but I think it is just more useful even if it's a bit unclear what it means for string arrays.
* Add `copy_to`, I made the `move_into()` now call a new `from_cudf()` to mirrow the `from_arrow()` helper and moved most error checks there just to not duplicate them.

So if/when you have a moment, may make sense to browse through once more.

seberg commented:
> I wonder if its easier when we just use eager allocations everywhere possible, preferring the ability to stream over temporary memory usage in tasks.

Yeah, although there are is important functionality where eager allocation is just not possible, so I am not sure it's worth transitioning to it wherever it is.  (But, although may well be if it helps legate with other things.)
